grpc_unary_attestation/src/server.rs

```diff
@@ -0,0 +1,81 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Server-side implementation of the bidirectional unary gRPC remote attestation handshake
+//! protocol.
+
+use crate::proto::{unary_session_server::UnarySession, UnaryRequest, UnaryResponse};
+use tonic;
+
+/// Trait for logging error messages.
+pub trait LogError {
+    fn log_error(&self, error: &str);
+}
+
+/// gRPC Attestation Service implementation.
+pub struct AttestationServer<F, L> {
+    /// PEM encoded X.509 certificate that signs TEE firmware key.
+    _tee_certificate: Vec<u8>,
+    /// Processes data from client requests and creates responses.
+    _request_handler: F,
+    /// Configuration information to provide to the client for the attestation step.
+    _additional_info: Vec<u8>,
+    /// Error logging function that is required for logging attestation protocol errors.
+    /// Errors are only logged on server side and are not sent to clients.
+    error_logger: L,
+}
+
+impl<F, S, L> AttestationServer<F, L>
+where
+    F: Send + Sync + Clone + FnOnce(Vec<u8>) -> S,
+    S: std::future::Future<Output = anyhow::Result<Vec<u8>>> + Send + Sync,
+    L: Send + Sync + Clone + LogError,
```

Comment:
Could these trait bounds be on the struct? I don't think it makes sense to create an AttestationServer struct in which error_logger is anything other than a LogError instance, right?

Reply:
I would think so. The thought occurred to me as well when reading `grpc_streaming_attestation/src/server.rs`, which this module mirrors. I'm not sure why it is done like that there (but there might be a reason). I felt it was probably best mirror that crate, but we could also update them both. 

WDYT?

Reply:
Yeah I think it's worth fixing in both, I don't think there was a good reason for that.